### PR TITLE
Link Control: Fix focus handlers in development mode

### LINKED ISSUE #62089: Link Control: Unable to edit text in development mode
The `LinkControl` popover steals the focus and prevents editing the text in development mode. See: #61943.

## Affected e2e tests:

* `editor/blocks/links.spec.js` - should allow editing text underneath popover when activated via mouse.
* `editor/various/block-bindings.spec.js` - should add empty button block when pressing enter.

## Steps to reproduce

* Checkout branch from #61943.
* Enable `SCRIPT_DEBUG` for non `wp-env` setups.
* Create a post.
* Add a paragraph and add a link to some selection.
* Try editing the linked text.

## Screencast

https://github.com/WordPress/gutenberg/assets/240569/d729d623-2ea8-4cad-98af-6fd862534923

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
Fixes #62089.
Related #61943.

PR fixes the `LinkControl` component's auto-focus handler and allows users to edit the linked text in development mode when the popover is rendered. 

## Why?
1. The initial value of  `isMounting` is true.
2. Effect runs and is set to false and returns early.
3. There's no cleanup, so the effect does nothing here.
4. Effect runs again; the `isMounting` is false, and focus logic is invoked.

## How?
Move the component's `isMounting` state synchronization into a separate effect and add a cleanup method.

## Testing Instructions
The following e2e test should pass in development mode, and CI checks should be green. See the original issue for enabling the dev mode.

```
npm run test:e2e -- editor/blocks/links.spec.js
npm run test:e2e -- editor/various/block-bindings.spec.js
```

### Testing Instructions for Keyboard
Same.

### CODE REVIEW

**Review by ellatrix — APPROVED** [review]
Ah, so it complains about the lack of cleanup function right?

**Comment by Mamaduka** [comment]
@ellatrix, yes, it's a similar scenario.

Here's the cycle:
1. The initial value of  `isMounting` is true.
2. Effect runs and is set to false and returns early.
3. There's no cleanup, so the effect does nothing here.
4. Effect runs again; the `isMounting` is false and focus logic is invoked.

The last two runs only happen in development mode to ensure that effect callbacks are pure. I used separate effect because effect also reacts to state change, and we don't want that to affect the `isMounted` flag.

TL;DR: If the component tracks the mounting state, it should also reset it on unmount.

**Review by jeryj — APPROVED** [review]
Looks good! Thanks for the fix.
